derive/src/pyclass.rs

```diff
@@ -1,91 +1,138 @@
 use super::rustpython_path_attr;
-use proc_macro2::{Span, TokenStream as TokenStream2};
+use proc_macro2::TokenStream as TokenStream2;
 use quote::quote;
+use std::collections::HashMap;
 use syn::{Attribute, AttributeArgs, Ident, ImplItem, Item, Lit, Meta, MethodSig, NestedMeta};
 
-enum MethodKind {
+#[derive(PartialEq, Clone, Copy)]
+enum ClassItemKind {
     Method,
-    Property,
+    PropertyGetter,
+    PropertySetter,
 }
 
-impl MethodKind {
-    fn to_ctx_constructor_fn(&self) -> Ident {
-        let f = match self {
-            MethodKind::Method => "new_rustfunc",
-            MethodKind::Property => "new_property",
-        };
-        Ident::new(f, Span::call_site())
-    }
+struct ClassItem {
+    item_name: Ident,
+    py_name: String,
+    kind: ClassItemKind,
 }
 
-struct Method {
-    fn_name: Ident,
-    py_name: String,
-    kind: MethodKind,
+fn meta_to_vec(meta: Meta) -> Result<Vec<NestedMeta>, Meta> {
+    match meta {
+        Meta::Word(_) => Ok(Vec::new()),
+        Meta::List(list) => Ok(list.nested.into_iter().collect()),
+        Meta::NameValue(_) => Err(meta),
+    }
 }
 
-impl Method {
-    fn from_syn(attrs: &mut Vec<Attribute>, sig: &MethodSig) -> Option<Method> {
-        let mut py_name = None;
-        let mut kind = MethodKind::Method;
-        let mut pymethod_to_remove = Vec::new();
-        let metas = attrs
+impl ClassItem {
+    fn extract_from_syn(attrs: &mut Vec<Attribute>, sig: &MethodSig) -> Option<ClassItem> {
+        let mut item = None;
+        let mut attr_idx = None;
+        for (i, meta) in attrs
             .iter()
+            .filter_map(|attr| attr.parse_meta().ok())
             .enumerate()
-            .filter_map(|(i, attr)| {
-                if attr.path.is_ident("pymethod") {
-                    let meta = attr.parse_meta().expect("Invalid attribute");
-                    // remove #[pymethod] because there's no actual proc macro
-                    // implementation for it
-                    pymethod_to_remove.push(i);
+        {
+            let name = meta.name();
+            if name == "pymethod" {
+                if item.is_some() {
+                    panic!("You can only have one #[py*] attribute on an impl item")
+                }
+                let nesteds = meta_to_vec(meta)
+                    .expect("#[pymethod = ...] must be a list, e.g. #[pymethod(...)]");
```

Comment:
I'm having trouble parsing this error message. Is `#[pymethod = ...]` actually allowed? If so, what does it do?

Reply:
No, it's not. The idea is that the error message will only be shown if somebody did write `#[pymethod = ""]`, so that would be a signal. Although, do we want to allow that? Basically as an alias to `#[pymethod (name = "")]`?

Reply:
I think I'd prefer not to allow that syntax in case we want additional options in the future. e.g. I've been considering the idea of `#[pymethod(constructor)]` that allows `new` methods to be more idiomatic rust.